cpp/src/io/parquet/experimental/hybrid_scan_preprocess.cu

```diff
@@ -0,0 +1,225 @@
+                                    cudf::detail::hostdevice_span<PageInfo> pages,
+                                    rmm::cuda_stream_view stream)
+{
+  CUDF_FUNC_RANGE();
+
+  std::vector<size_t> host_chunk_page_counts(chunks.size() + 1);
+  std::transform(
+    chunks.host_begin(), chunks.host_end(), host_chunk_page_counts.begin(), [](auto const& chunk) {
+      return chunk.num_dict_pages;
+    });
+  host_chunk_page_counts[chunks.size()] = 0;
+
+  auto chunk_page_counts = cudf::detail::make_device_uvector_async(
+    host_chunk_page_counts, stream, cudf::get_current_device_resource_ref());
+
+  thrust::exclusive_scan(rmm::exec_policy_nosync(stream),
+                         chunk_page_counts.begin(),
+                         chunk_page_counts.end(),
+                         chunk_page_counts.begin(),
+                         size_t{0},
+                         thrust::plus<size_t>{});
+
+  rmm::device_uvector<chunk_page_info> d_chunk_page_info(chunks.size(), stream);
+
+  thrust::for_each(rmm::exec_policy_nosync(stream),
+                   thrust::counting_iterator<cuda::std::size_t>(0),
+                   thrust::counting_iterator(chunks.size()),
+                   [cpi               = d_chunk_page_info.begin(),
+                    chunk_page_counts = chunk_page_counts.begin(),
+                    pages             = pages.device_begin()] __device__(size_t i) {
+                     cpi[i].pages = &pages[chunk_page_counts[i]];
+                   });
+
+  parquet::kernel_error error_code(stream);
+
+  DecodePageHeaders(
+    chunks.device_begin(), d_chunk_page_info.begin(), chunks.size(), error_code.data(), stream);
+
+  if (auto const error = error_code.value_sync(stream); error != 0) {
+    CUDF_FAIL("Parquet header parsing failed with code(s) " +
+              parquet::kernel_error::to_string(error));
+  }
+
+  // Setup dictionary page for each chunk
+  thrust::for_each(rmm::exec_policy_nosync(stream),
+                   pages.device_begin(),
+                   pages.device_end(),
+                   [chunks = chunks.device_begin()] __device__(PageInfo const& p) {
+                     if (p.flags & parquet::detail::PAGEINFO_FLAGS_DICTIONARY) {
+                       chunks[p.chunk_idx].dict_page = &p;
+                     }
+                   });
+
+  pages.device_to_host_async(stream);
+  chunks.device_to_host_async(stream);
+  stream.synchronize();
+}
+
+}  // namespace
+
+std::tuple<bool,
+           cudf::detail::hostdevice_vector<ColumnChunkDesc>,
+           cudf::detail::hostdevice_vector<PageInfo>>
+hybrid_scan_reader_impl::prepare_dictionaries(
+  cudf::host_span<std::vector<size_type> const> row_group_indices,
+  cudf::host_span<rmm::device_buffer> dictionary_page_data,
+  cudf::host_span<int const> dictionary_col_schemas,
+  parquet_reader_options const& options,
+  rmm::cuda_stream_view stream)
+{
+  // Create row group information for the input row group indices
+  auto const row_groups_info =
```

Comment:
This function is is modified from `reader::impl::create_global_chunk_info()` function in `reader_impl_chunking.cu`